### ISSUE TITLE
False positive `E0308` involving closure and `impl FnMut`

### ISSUE THREAD

**Issue by ronnodas** [rust-lang/rust-analyzer#21819]
**rust-analyzer version**: rust-analyzer version: 0.3.2819-standalone

**rustc version**: rustc 1.94.0 (4a4ef493e 2026-03-02)

**editor or extension**: VS Code, extension version 0.3.2819, running on WSL

**relevant settings**: raised with experimental diagnostics either enabled or disabled.

**code snippet to reproduce**:
This is the code that I encountered it on, I haven't tried minimizing it. Compiles fine but RA generates an E0308 diagnostic: "expected Map<Self, impl FnMut({unknown}) -> Option<U>>, found Map<Self, {closure#253867}>"

```rust
trait IteratorExt: Iterator {
    fn map_some<F: FnMut(T) -> U, T, U>(
        self,
        mut f: F,
    ) -> std::iter::Map<Self, impl FnMut(Self::Item) -> Option<U>>
    where
        Self: Iterator<Item = Option<T>> + Sized,
    {
        self.map(move |x| x.map(&mut f))
    }
}
```

**Comment by ChayimFriedman2** [comment]
I can't reproduce on prerelease r-a.

**Comment by ronnodas** [comment]
Can confirm, doesn't seem to occur on 0.4.2826.
